### PR TITLE
Add cache to sequencer

### PR DESCRIPTION
**Pull request by samfrench**
Add cache to sequencer

https://jira.dev.bbc.co.uk/browse/CONNPOL-2821

This is the second part of the DynamoDB caching strategy.

### CODE REVIEW

**Comment by samfrench** [comment]
@BBCMarkMcDonnell, this is the second caching PR, so we can have the cache in front of both of our Dynamo tables.

I updated the specs to RSpec 3 syntax, and did a bit of a tidy up, then added the cache and fixed the specs.

I created an instance of the cache in the sequencer class, but could have created it in the sequencer/sequencer class, but decided to base it on the sequence table one.

I added a test to the sequencer spec file, but I think it would be worth splitting it into a cache_spec file, to keep them separate? Then I could provide the cache as a double, the same as the sequence table. As the sequence table is a double, it would be worth having a table_spec file, to test the logic in the table class, as we assume it will work as expected. Any thoughts on the specs, or is one file fine?

File: lib/alephant/sequencer.rb [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 require "alephant/sequencer/sequencer"
 require "alephant/sequencer/sequence_table"
+require "alephant/sequencer/sequence_cache"
 
 module Alephant
   module Sequencer
     @@sequence_tables = {}
 
-    def self.create(table_name, ident, jsonpath = nil, keep_all = true)
+    def self.create(table_name, ident, jsonpath = nil, keep_all = true, config = {})
```
Comment (Integralist) [inline lib/alephant/sequencer.rb]: This method signature is starting to get a bit long winded. Be good to see if we can refactor this within this PR (e.g. consolidate the 2nd, 3rd, 4th and the new 5th argument)

So it looks something like:

```rb
self.create(table_name, opts)
```

Where `opts` would be a hash like:

```rb
{
  :ident    => foo,
  :jsonpath => bar,
  :keep_all => baz,
  :config   => {}
}
```

File: lib/alephant/sequencer.rb [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
       @@sequence_tables[table_name] ||= SequenceTable.new(table_name)
-      Sequencer.new(@@sequence_tables[table_name], ident, jsonpath, keep_all)
+      @cache ||= SequenceCache.new(config)
+      Sequencer.new(@@sequence_tables[table_name], ident, jsonpath, keep_all, @cache)
```
Comment (Integralist) [inline lib/alephant/sequencer.rb]: This method signature is starting to get a bit unwieldy too, so a similar refactor would be nice

File: lib/alephant/sequencer/sequence_cache.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+      def get(key, &block)
+        begin
+          versioned_key = versioned key
+          result = @client.get versioned_key
+          logger.info "Alephant::SequenceCache#get key: #{versioned_key} - #{result ? 'hit' : 'miss'}"
+          logger.metric "GetKeyMiss" unless result
+          result ? result : set(key, block.call)
+        rescue StandardError => e
+          block.call if block_given?
```
Comment (Integralist) [inline lib/alephant/sequencer/sequence_cache.rb]: We check `if block_given?` here (when we error), and yet we don't check for the blocks existence from within the `begin`'s scope? 

If we were concerned that a block wasn't going to be provided, then I'd expect we'd need the same check when calling `set`

**Comment by Integralist** [comment]
Hi @samfrench I've made some minor comments for you to review, otherwise this PR seems fine to me.

In response to your question about splitting up the specs, that sounds fine to me. I'd prefer to inject as many as the dependencies as possible to ensure the tests can become as isolated as possible :+1:

File: lib/alephant/sequencer/sequence_cache.rb (hunk 3 above)
Comment (samfrench) [inline lib/alephant/sequencer/sequence_cache.rb]: I did see this the other day, and I assumed that the error would be caught by the rescue if a block wasn't passed in. At the moment, the get is only called within the Sequencer/Sequencer class and a block if provided.

Is it worth removing the `block_given?` check, as the block is provided, and I can't see a case where we would add functionality and not provide a block if the item wasn't in the cache.

Comment (Integralist) [inline lib/alephant/sequencer/sequence_cache.rb]: Yeah I would remove the `block_given?` actually. It's only _ever_ provided by us in `lib/alephant/sequencer/sequencer.rb` as far as I can see.

**Comment by samfrench** [comment]
@Integralist, I have made amendments based on your feedback.

I will leave the tests as they are at the moment, as that could be time-consuming to split them up, when they aren't currently posing an issue. I'll look at updating the initialisation of the Sequencer and Lookup in the other Alephant gems, so we can actually use the cache.

**Comment by Integralist** [comment]
:cake:

**Pull request merged**